Replace 'go-gomail/gomail' with 'wneessen/go-mail'

wneessen opened the pull request:
This PR addresses https://github.com/offen/offen/issues/817. It replaces the legacy and not maintained 'go-gomail/gomail' package with 'wneessen/go-mail'.

* Generally this PR changes the signature of the `mailer` constructor to also return an error, since 'wneessen/go-mail' requires this. 
* Additionally it adds `Authtype` as config option. Be default this is set to `LOGIN`, but can be either of `LOGIN`, `PLAIN` or `CRAM-MD5`.
* Since 'wneessen/go-mail' supports local sendmail delivery out of the box, some of the pervious sendmail code has been thrown away.

Commit:
Change signature for mail feature

The mailer constructors for SMTP, local, and sendmail have an update signature and now return an error, in addition to the created Mailer object. This change is a preparation for the introduction of wneessen/go-mail, which returns an error with the New() method.

Commit:
Replace 'go-gomail/gomail' with 'wneessen/go-mail'

This commit replaces the deprecated 'go-gomail/gomail' package with 'wneessen/go-mail'. Additionally it adds `Authtype` as config option. Be default this is set to `LOGIN`, but can be either of `LOGIN`, `PLAIN` or `CRAM-MD5`. Since 'wneessen/go-mail' supports local sendmail delivery out of the box, some of the pervious sendmail code has been thrown away.

m90 commented on server/config/definition_unix.go:
```diff
@@ -36,6 +36,7 @@ type Config struct {
 	}
 	Secret Bytes
 	SMTP   struct {
+		Authtype string `default:"LOGIN"`
```
Could you add the possible options and their implications to the documentation site still? https://github.com/offen/offen/blob/development/docs/docs/running-offen/configuring-the-application.md is probably the location for that.

wneessen replied:
Added some documentation. Does this suffice or should I go more into detail?

m90 replied:
I think that's fine. Thank you.

m90 commented on server/mailer/smtpmailer/smtpmailer.go:
```diff
@@ -4,25 +4,51 @@
+	case "cram-md5":
+		c.SetSMTPAuth(mail.SMTPAuthCramMD5)
+	default:
+		return nil, errors.New("configured SMTP auth type is not supported")
```
```suggestion
		return nil, fmt.Errorf("configured SMTP auth type %s is not supported", authtype)
```

Commit:
Add 'noauth' case to SMTP authentication type

The 'noauth' case is now added to the SMTP authentication type in the SMTPMailer. Since in some enterprise environments, SMTP auth might not be supported on local SMTP relays.

Commit:
Refine error messaging for unsupported SMTP auth types

The error message for unsupported SMTP authentication types has been updated to include the specific type that is not supported. Previously, a static message was returned regardless of the unsupported auth type. Now, if an unsupported auth type is encountered, it'll be clearly identified in the error message, improving the system's usability and maintainability.

Commit:
Add SMTP authentication type configuration details

This change includes a new section in the configuring-the-application.md document for the OFFEN_SMTP_AUTHTYPE. This new configuration allows users to specify the authentication type used for sending transactional emails, with supported types indicated.

m90 commented on docs/docs/running-offen/configuring-the-application.md:
```diff
@@ -211,6 +211,13 @@ Default value `no-reply@offen.dev`.
+
+Default value `LOGIN`.
+
+The SMTP authentication type used when sending transactional email. Supported types are: `LOGIN`, `PLAIN`, `CRAM-MD5`, `NOAUTH`.
```
As the code now rejects `NOAUTH` I wonder if we should omit it here? Any specific reason you included it in the list?

wneessen replied:
I've snuck in support for "noauth" with commit https://github.com/offen/offen/pull/827/commits/cdcb24df615089e667650db4e8dddc95b59b3a19. I thought it might be useful in some enterprise settings.

m90 replied:
Sorry, I forgot that Go does not fallthrough to default, but this means it instead does nothing. Nevermind.

m90 reviewed (approved):
Brilliant, thank you. I'll let you know when the next release goes out and this can be used.
